GraphQL/DataLoader/CachedContentLoader.php

```diff
@@ -0,0 +1,49 @@
+<?php
+namespace BD\EzPlatformGraphQLBundle\GraphQL\DataLoader;
+
+use BD\EzPlatformGraphQLBundle\GraphQL\DataLoader\Exception\NotFoundException\NotFoundException;
+use eZ\Publish\API\Repository\Exceptions as ApiException;
+use eZ\Publish\API\Repository\SearchService;
+use eZ\Publish\API\Repository\Values\Content\Content;
+use eZ\Publish\API\Repository\Values\Content\Query;
+use eZ\Publish\API\Repository\Values\Content\Search\SearchHit;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+
+class CachedContentLoader implements ContentLoader
+{
+    /**
+     * @var ContentLoader
+     */
+    private $innerLoader;
+
+    private $loadedItems = [];
+
+    public function __construct(ContentLoader $innerLoader)
+    {
+        $this->innerLoader = $innerLoader;
+    }
+
+    public function find(Query $query): array
+    {
+        $items = $this->innerLoader->find($query);
+
+        foreach ($items as $item) {
+            $this->loadedItems[$item->id] = $item;
+        }
+
+        return $items;
+    }
+
+    public function findSingle(Criterion $filter): Content
+    {
+        $contentId = $filter->value[0];
+        if ($filter instanceof Criterion\ContentId && isset($this->loadedItems[$contentId])) {
```

Comment:
if this is often the case you can also map top load methods so they are cached on API/SPI side alos when yet in local cache here.


aka:
```
if ($filter instanceof Criterion\ContentId) {
    if (isset($this->loadedItems[$contentId])) {
        return $this->loadedItems[$contentId];
    } else {
        // As of EZP-29899 (7.4.0, and upcoming 7.3.5, 6.13.6, 6.7.10, 5.4.14) we don't need to specify current version
        return $this->loadedItems[$contentId] = $this->contentService->load($contentId);
    }
```


You can use the same trick in find(), however maybe only if you detect* presence of bulk loading api's:
- `loadContentInfoList` : Coming as part of https://github.com/ezsystems/ezpublish-kernel/pull/2529
- `loadContentListByContentInfo`: Added as part of https://github.com/ezsystems/ezpublish-kernel/commit/2a3684893a2714a8c5c07ea1f4f5bffa3fb2f94c

_(or we could expose a simplified `loadContentList($ids)` also to skip the round trips)_

_Side: You should btw have the bulk/multi-load api's in mind when doing design here, design of ContentTypeLoader for instance will explicitly disallow you from being able to take advantage of the recent method to load several content types at once on ContentTypeService._

* _Assuming you aim to make this support several LTS kernel versions to assure broad reach._

Reply:
Thank you for your feedback, @andrerom, it helps. I'll change single content item loading to what you suggested, as it makes no sense to use the SearchService to load a single item. 

I'm actually not entirely set about the load method itself, though: it was either a unique method, with a flexible argument (in that case, a Criterion), or one method per supported argument. Would you by any chance have arguments against / in favour of one or the other ?

> design of ContentTypeLoader for instance

Actually, this one shouldn't have made it, I'll remove it from the PR for the time being. But I'll keep your remarks in mind for the future design of it.

Reply:
It's about `findSingle()` above: I'm not sure yet if I want `findSingle()` or `loadById()`, `loadByRemoteId()`, `loadByLocationId()`...

Reply:
Oh, no strong opinion there, but of course if consumer is your own code here:
1. load and findSingle has same semantics, it throws if missing. But findSignal adds unneeded overhead if you anyway end up mapping it to load method(s)
2. The bulk/multi load suggestion still applies, if there is a need to load several, allow it to not force un-needed round trips to backend _(do backfire profiling on server setup where redis is on network to see what I mean)_

Reply:
> The bulk/multi load suggestion still applies, if there is a need to load several, allow it to not force un-needed round trips to backend 

I have read your suggestions several times, and while I see the idea of those methods, I'm not sure in which case they should apply here. I'll keep it in mind for the `ContentTypeLoader`, but in regards to the Content one, it is not clear. Should I change my usage of `SearchService::findContent()` to take advantage of this or not ?

> (do backfire profiling on server setup where redis is on network to see what I mean)
Actually, I have, on Platform.sh.

Reply:
Let's find an occasion to discuss, I might need to better understand what you are trying to do